File: src/YowThi.Erp.Application/Processing/ProcessingExecutionOptions.cs
using YowThi.Erp.Domain.ProcessingConfiguration;

namespace YowThi.Erp.Application.Processing;

public sealed record ProcessingExecutionOptionsQuery(
    string Locale,
    string? Search,
    int Offset,
    int Limit);

public sealed record ProcessingExecutionOptionPage<T>(
    IReadOnlyList<T> Items,
    int? NextOffset);

public sealed record ProcessingEmployeeOption(
    Guid Id,
    string DisplayName);

public sealed record ProcessingBatchOption(
    Guid Id,
    DateOnly ProcurementDate,
    Guid ProcurementProductId,
    string ProcurementProductDisplayName,
    Guid ProcessingRouteVersionId);

public sealed record ProcessingModuleOption(
    Guid Id,
    string DisplayName,
    ProcessingExecutionMode ExecutionMode,
    Guid? InputProcessMaterialId,
    bool? InputUsesContainer,
    Guid? InputContainerId,
    int? DefaultInputContainerCount);

public sealed record ProcessingBatchModuleOptions(
    Guid ProcurementBatchId,
    Guid ProcessingRouteVersionId,
    ProcessingExecutionOptionPage<ProcessingModuleOption> Modules);

public sealed record ProcessingSupplierOption(
    Guid Id,
    string DisplayName);

public sealed record ProcessingStorageLocationOption(
    Guid Id,
    string DisplayName,
    string? Code,
    Guid WarehouseId);

public sealed record ProcessingInputStorageLocationOptions(
    Guid ProcurementBatchId,
    Guid ProcessingModuleId,
    Guid? AutoSelectionLocationId,
    ProcessingExecutionOptionPage<ProcessingStorageLocationOption> Locations);

public sealed record ProcessingModuleOutputOption(
    Guid Id,
    int OutputSequence,
    ProcessingOutputKind OutputKind,
    Guid? TargetId,
    string? TargetDisplayName,
    bool TargetAvailable,
    bool UsesContainer,
    Guid? ContainerId,
    int? DefaultContainerCount,
    decimal? PackagingWeight,
    Guid? DefaultStorageLocationId,
    bool DefaultStorageLocationAvailable,
    decimal DefaultWageRate);

public sealed record ProcessingModuleOutputOptions(
    Guid ProcurementBatchId,
    Guid ProcessingModuleId,
    IReadOnlyList<ProcessingModuleOutputOption> Outputs);

public interface IProcessingExecutionOptionsReader
{
    ValueTask<ProcessingExecutionOptionPage<ProcessingEmployeeOption>> GetEmployeesAsync(
        ProcessingExecutionOptionsQuery query,
        CancellationToken cancellationToken);

    ValueTask<ProcessingExecutionOptionPage<ProcessingBatchOption>> GetBatchesAsync(
        ProcessingExecutionOptionsQuery query,
        CancellationToken cancellationToken);

    ValueTask<ProcessingBatchModuleOptions?> GetModulesAsync(
        Guid procurementBatchId,
        ProcessingExecutionOptionsQuery query,
        CancellationToken cancellationToken);

    ValueTask<ProcessingExecutionOptionPage<ProcessingSupplierOption>> GetSuppliersAsync(
        ProcessingExecutionOptionsQuery query,
        CancellationToken cancellationToken);

    ValueTask<ProcessingInputStorageLocationOptions?> GetInputStorageLocationsAsync(
        Guid procurementBatchId,
        Guid processingModuleId,
        ProcessingExecutionOptionsQuery query,
        CancellationToken cancellationToken);

    ValueTask<ProcessingModuleOutputOptions?> GetModuleOutputsAsync(
        Guid procurementBatchId,
        Guid processingModuleId,
        string locale,
        CancellationToken cancellationToken);

    ValueTask<ProcessingExecutionOptionPage<ProcessingStorageLocationOption>> GetStorageLocationsAsync(
        ProcessingExecutionOptionsQuery query,
        CancellationToken cancellationToken);
}

File: src/YowThi.Erp.Web/src/features/processing/processingExecutionOptions.ts
import { getApiJson } from '../../app/api/apiTransport';
import type { OperationalLocale } from '../../app/i18n/locale';

export type ProcessingExecutionMode = 'SOURCE_TRACKED' | 'POOLED_OUTPUT' | 'FINAL_PACKAGING';
export type ProcessingOutputKind = 'PROCESS_MATERIAL' | 'SALES_PRODUCT';

export interface ProcessingEmployeeOption {
  id: string;
  displayName: string;
}

export interface ProcessingEmployeeOptionsResponse {
  items: ProcessingEmployeeOption[];
  nextCursor: string | null;
}

export interface ProcessingBatchOption {
  id: string;
  procurementDate: string;
  procurementProductId: string;
  procurementProductDisplayName: string;
  processingRouteVersionId: string;
}

export interface ProcessingBatchOptionsResponse {
  items: ProcessingBatchOption[];
  nextCursor: string | null;
}

export interface ProcessingModuleOption {
  id: string;
  displayName: string;
  executionMode: ProcessingExecutionMode;
  inputProcessMaterialId: string | null;
  inputUsesContainer: boolean | null;
  inputContainerId: string | null;
  defaultInputContainerCount: number | null;
}

export interface ProcessingBatchModuleOptionsResponse {
  procurementBatchId: string;
  processingRouteVersionId: string;
  items: ProcessingModuleOption[];
  nextCursor: string | null;
}

export interface ProcessingSupplierOption {
  id: string;
  displayName: string;
}

export interface ProcessingSupplierOptionsResponse {
  items: ProcessingSupplierOption[];
  nextCursor: string | null;
}

export interface ProcessingStorageLocationOption {
  id: string;
  displayName: string;
  code: string | null;
  warehouseId: string;
}

export interface ProcessingStorageLocationOptionsResponse {
  items: ProcessingStorageLocationOption[];
  nextCursor: string | null;
}

export interface ProcessingInputStorageLocationOptionsResponse {
  procurementBatchId: string;
  processingModuleId: string;
  autoSelectionLocationId: string | null;
  items: ProcessingStorageLocationOption[];
  nextCursor: string | null;
}

export interface ProcessingModuleOutputOption {
  id: string;
  outputSequence: number;
  outputKind: ProcessingOutputKind;
  targetId: string | null;
  targetDisplayName: string | null;
  targetAvailable: boolean;
  usesContainer: boolean;
  containerId: string | null;
  defaultContainerCount: number | null;
  packagingWeight: number | null;
  defaultStorageLocationId: string | null;
  defaultStorageLocationAvailable: boolean;
  defaultWageRate: number;
}

export interface ProcessingModuleOutputOptionsResponse {
  procurementBatchId: string;
  processingModuleId: string;
  outputs: ProcessingModuleOutputOption[];
}

interface OptionQuery {
  locale: OperationalLocale;
  search?: string;
  cursor?: string | null;
  limit?: number;
  signal?: AbortSignal;
}

export async function listProcessingEmployeeOptions(
  query: OptionQuery,
): Promise<ProcessingEmployeeOptionsResponse> {
  return getJson('/api/v1/processing/execution-options/employees', query);
}

export async function listProcessingBatchOptions(
  query: OptionQuery,
): Promise<ProcessingBatchOptionsResponse> {
  return getJson('/api/v1/processing/execution-options/batches', query);
}

export async function listProcessingModuleOptions(
  procurementBatchId: string,
  query: OptionQuery,
): Promise<ProcessingBatchModuleOptionsResponse> {
  return getJson(
    '/api/v1/processing/execution-options/modules',
    query,
    { procurementBatchId },
  );
}

export async function listProcessingSupplierOptions(
  query: OptionQuery,
): Promise<ProcessingSupplierOptionsResponse> {
  return getJson('/api/v1/processing/execution-options/suppliers', query);
}

export async function listProcessingInputStorageLocationOptions(
  procurementBatchId: string,
  processingModuleId: string,
  query: OptionQuery,
): Promise<ProcessingInputStorageLocationOptionsResponse> {
  return getJson(
    '/api/v1/processing/execution-options/input-storage-locations',
    query,
    { procurementBatchId, processingModuleId },
  );
}

export async function getProcessingModuleOutputOptions(
  procurementBatchId: string,
  processingModuleId: string,
  locale: OperationalLocale,
  signal?: AbortSignal,
): Promise<ProcessingModuleOutputOptionsResponse> {
  return getJson(
    '/api/v1/processing/execution-options/module-outputs',
    { locale, signal },
    { procurementBatchId, processingModuleId },
  );
}

export async function listProcessingStorageLocationOptions(
  query: OptionQuery,
): Promise<ProcessingStorageLocationOptionsResponse> {
  return getJson('/api/v1/processing/execution-options/storage-locations', query);
}

async function getJson<T>(
  path: string,
  query: OptionQuery,
  requiredParameters: Record<string, string> = {},
): Promise<T> {
  const parameters = new URLSearchParams(requiredParameters);
  const search = query.search?.trim();

  if (search) {
    parameters.set('search', search);
  }

  if (query.cursor) {
    parameters.set('cursor', query.cursor);
  }

  if (query.limit !== undefined) {
    parameters.set('limit', String(query.limit));
  }

  const suffix = parameters.size === 0 ? '' : `?${parameters.toString()}`;
  return getApiJson<T>(`${path}${suffix}`, {
    locale: query.locale,
    signal: query.signal,
  });
}
